### ISSUE TITLE
Events running on nested fields

### ISSUE THREAD

**Issue by willrbc** [jagi/meteor-astronomy#214]
Hi, we are running into an issue with events running on a nested class calls a method.  The nested class is a 'Price' class with 'cost', 'price' and 'markup' fields.

The the idea is the have the 'markup' field set when cost or price are changed.

The event appears to run and this.set('markup', calculateMarkup(this.cost, this.price)) runs successfully.

The strange this is that the markup field doesn't change. If you call console.log(this) immediately after it is changed then object returned still has markup as 'null' however if you console.log(this.markup) you get the correct amount for markup.

```javascript
Prices = Astro.Class({
  name: 'Prices',
  fields: {
    cost: {
      type: 'number'
    },
    price: {
      type: 'number'
    },
    markup: {
      type: 'number'
    }
  },
  events: {
    afterChange: function (e) {
      if (e.data.fieldName === 'price') {
        this.calculateMarkup();
      }      
    }
  },
  methods: {
    calculateMarkup: function () {
      if (this.cost && this.price) {
        console.log("calculateMarkup: " + this.cost * this.price);
        this.set('markup', Library.calculateMarkup(this.cost, this.price));
        console.log(this);  //  markup is null
        console.log(this.markup); // markup is correct!
      }
    }
});
```

The above class was nested inside a Product class, when I moved the cost fields into the top level of the Product class, all events ran correctly and behaviour was as expected.

Is there a known issue with nested events and is there likely to be a fix?  This library is very powerful for my project, but I really need multi layered nested events being triggered and functioning correctly.

Many thanks

**Comment by jagi** [comment]
Could you create a reproduction repository? I don't have time to create it by myself and it would be easier to investigate a problem. There are some limitations of events if it goes about nested fields but don't know if it applies here.

**Comment by willrbc** [comment]
Hi Jagi

https://github.com/willrbc/nested-events

I've been experimenting using your examples repository.  There is now a Products collection with nested 'bundle products'

The price is set at both top level and in the nested environment,  

I've created a simple button that you can press to modify the top level price as well as nested price and see the differing results.

The events are triggered, bit their results are not written.

Many thanks

Will

**Comment by jagi** [comment]
I don't know what's going on in your reproduction. I've spent 30 minutes trying to understand it. Please create as minimal reproduction as possible. And please don't use an example project without removing everything that is not necessary.

For now, I can only say that when you are accessing `this.markup` and it's value is not an object then you just see the result on the console. But when you console log `this` then it's an object reference which can change during the time. And when you're accessing it from the console it can be in a different state then it was during the console.log. But it's quite obvious if you know how JS manages data. So it's not a bug. The problem is probably somewhere else. But as I said I can't spend hours on investigating problem. Try splitting it apart and just reproduce minimal problem. Just leave you products, prices problem aside and try preparing reproduction that replicates the problem.

**Comment by willrbc** [comment]
Hi Jagi,

I've created a simple repo from scratch and I don't seem to be having the same problem.  I will work backwards from the above Repo towards my new minimal one and see where the difference occurs.

**Comment by jagi** [comment]
OK great, so try replicating a bug and I will look at your repo in a moment.

**Comment by willrbc** [comment]
Okay,  just to confirm the second repo (astroTest) is working,  the other one still isn't,  I will continue to simplify the other one and send another commit

**Comment by jagi** [comment]
Ok good

**Comment by willrbc** [comment]
Morning Jagi

I've think I've discovered the issue why one worked and not the other. (They both now work).

So on my 'nested-events' repo, the price Class had a function triggered 'afterChange' that would set the markup field as a calculation of cost and price when either of those fields was being edited.

I created a simple way of setting new random values of cost and price on a button press.

While doing a lot of logging on the 'astroTest' repo, I discovered that, effectively, each field was being changed everytime I ran .save()  even if it was no different.  On my first repo with prices the 'markup' field was saved as it's original value by running .save(), this would overwrite the calculated value set when the cost or price field was changed.  I managed to fix this by changing the order on which the fields are declared.  By declaring markup first in the fields object, its original value would be saved before the cost or price fields could trigger it to be changed, as that now happens afterwards, the new correct markup figure sticks.

It might also be that I'm not doing the getting and setting correctly on my prices repo,  loading the entire object form the database and setting it may be what causes the double writing.

I hope this makes sense, I'm not the best at explaining this sort of thing.

Many thanks for all your help and work on Astronomy

**Comment by willrbc** [comment]
To clarify my earlier post.

The reason I initially thought that the issue was to do with Nesting is because when I first built my model, the behavour of the top level price was as expected, but the lower level prices were not.

When I came back to it,  the behaviour did not work regardless of whether it was top level or nested.

I now believe it had more to do with the order in which fields are modified, and perhaps inconsistencies were be caused by the asynchronous nature of the javascript?

**Comment by jagi** [comment]
Ok I know what's going on. There is a design error (I would not call it bug) in when the `change` event is triggered. Let's take such the line `Items.findOne()`. The change event will be called even on fetching document from a database. But it's wrong. It should be triggered on calling the `set()`, `push()` etc. methods. However, I can't change it now because it could break somebody's code. This change would not be backward compatible. In Astronomy 2.0 it will be changed. There will be no `beforeSet`, `beforePush` etc. event. Instead there will be only `beforeChange` and `afterChange` events. They will be called when you write `doc.save()`. These event will tell you what field has changed and what is the current and previous field's value.

There is one hack to distinguish if the `before/afterChange` event was triggered during doc initialisation or actual change.

```js
Item.extend({
  events: {
    beforeChange: function(e) {
      if (e.target._isNew !== undefined) {
        console.log('beforeChange');
      }
    }
  }
});
```

**Comment by willrbc** [comment]
Yes that helps a lot, thank you.  Now that I've understood the issue and got it sort of working, I'm going to build up the complexity until its the model I require and keep testing.

Will update you.

Is there a roadmap / timescale for 2.0?

**Comment by willrbc** [comment]
Yes it works!

One thing I discovered is the difference between setting:

p.set("top.mid.level", value)

and

p.top.mid.set("level", value)

the 2nd one sets a value for _isNew  whereas the first one leaves it undefined.

Thanks for your help,  this datastructure is now working great for us and we can crack on with development!

**Comment by jagi** [comment]
In fact the `_isNew` property should not exist in the nested document. So it's actually isn't important. The `set()` method is not setting the `_isNew` property. It's undefined on document initialisation. And it's set to `true` after initialisation or to `false` if it's a document fetched from the collection.

**Comment by jagi** [comment]
I hope to release 2.0.0-rc.1 version until the end of the year but there are many variables that can influence the release date.

**Comment by crussi** [comment]
Hi Jagi,
I left a question on your documentation chat in regards to "relations".
I'm not sure the best way to reach you.  You're project is awesome and I
seem to be stuck despite much studying on my end.
Chris

**Comment by willrbc** [comment]
Okay, so although we are relying on this _isNew property at the moment to distinguish between the genuine triggering of 'afterChange' we won't need to after the bug you mentioned is fixed for 2.0 because each triggering of 'afterChange' will be genuine.

As we are at the very beginning of our project it would be good to keep an eye on these changes for 2.0 and if possible use 2.0 ASAP.

Many thanks Jagi

**Comment by jagi** [comment]
Yes I think it will be a big improvement when switching to 2.0. I have to hurry up with this because in our team we will be switching to 2.0 for one project so it's important to deliver it ASAP.
